### PR TITLE
[bug fix] using snr gamma and prior preservation loss in the dreambooth lora sdxl training scripts

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR _fixes_ bug when using both min SNR gamma and prior preservation loss in the`train_dreambooth_lora_sdxl.py` and `train_dreambooth_lora_sdxl_advanced.py` training scripts:

**Reproducing the error:**
when enabling both `--snr_gamma` and `--with_prior_preservation`, the following line errors:
`loss = loss.mean(dim=list(range(1, len(loss.shape)))) * mse_loss_weights`
Due to a shape mismatch between the instance loss (`loss`) and the `mse_loss_weights`.

I think the issue is the shape of timesteps - when we calculate snr gamma we use timesteps sampled for _all_ images in the batch (i.e. both instance and class images) rather than only instance images.
chunking the timesteps used for snr gamma calculation in 2 seem to solve the bug - however not sure if it 100% makes sense logically. 
@sayakpaul @apolinario WDYT?

### CODE REVIEW

File: examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py [hunk 1]
```diff
@@ -1799,9 +1799,17 @@ def compute_text_embeddings(prompt, text_encoders, tokenizers):
...
+                        # if we're using prior preservation, we calc snr for instance loss only -
+                        # and hence only need timesteps corresponding to instance images
+                        snr_timesteps, _ = torch.chunk(timesteps, 2, dim=0)
+                    else:
+                        snr_timesteps = timesteps
+
+                    snr = compute_snr(noise_scheduler, snr_timesteps)
                     base_weight = (
-                        torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(dim=1)[0] / snr
+                        torch.stack([snr, args.snr_gamma * torch.ones_like(snr_timesteps)], dim=1).min(dim=1)[0] / snr
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1812]: Shouldn't this affect the regular SDXL LoRA trainer too? Have we verified that? I am also not too sure if formulation-wise this is the correct thing to do i.e., when prior-preservation is done how should Min-SNR be computed. Your fix seems reasonable but I am honestly not very sure about the validity.

Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1812]: > Shouldn't this affect the regular SDXL LoRA trainer too? Have we verified that?

I believe it does, since it's independent of the pivotal tuning training, that's why I pr'd the change for both trainers.

Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1812]: Regarding the validity, I'm not very sure as well, my approach was in part based on the way we handled the use of both prior preservation loss and snr gamma in the [dreambooth training script](https://github.com/huggingface/diffusers/blob/4c483deb9077ef733d568ff0ae47233c35efd039/examples/dreambooth/train_dreambooth.py#L1280C17-L1305C1) 
where the snr gamma calculation is also seemingly computed only with regards to the instance loss (?). 
Maybe this sheds some light?

Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_sdxl_advanced.py:1812]: It does. Thanks. Let's proceed with that, but also, verify some results with experiments?

**Comment by sayakpaul** [comment]
@linoytsaban I am happy to merge if you can confirm that it doesn't negatively affect the quality.

**Comment by linoytsaban** [comment]
> @linoytsaban I am happy to merge if you can confirm that it doesn't negatively affect the quality.

Sorry missed this! In my experiments I didn’t see a quality degradation.  
Maybe we can merge it to the advanced script and keep investigating it mathematically / ask the community before we merge into the canonical?

**Comment by sayakpaul** [comment]
Fine by me. Let's make the changea accordingly then.

**Comment by linoytsaban** [comment]
Done :)
